### ISSUE TITLE
prefer-const detect unnecessary let

### ISSUE THREAD

**Issue by graingert** [eslint/eslint#9410]
**Please describe what the rule should do:**
Detect uses of let where reassignments and accesses happen only in the same function, Eg not in a nested function or loop.

If a variable is only reassigned and accessed inside one function, then the value must have stabilized by the point of the function's return, therefore it can always be replaced with a const. (While this also applies to loops, the developer would have to switch to using reduce or recursion)

**What category of rule is this? (place an "X" next to just one item)**

[ x] Enforces code style
[ ] Warns about a potential error
[x ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// bad
function foo() {
  let itWorks = false;
  try {
    checkIfItWorks();
    // only mutated/accessed inside function foo
    itWorks = true;
  } catch {
    // only mutated/accessed inside function foo
    itWorks = false;
  }
  ...
}

// good
function foo() {
  const itWorks = (() => 
    try {
      checkIfItWorks();
      return true;
    } catch {
      return false;
    }
  })();
   ...
}

// bad
function foo() {
  let result = getResult();
  if (!result) {
    // only mutated/accessed inside function foo
    result = getDefaultResult();
  }

  ...;
}

// good
function foo() {
  const result = getResult() || getDefaultResult();
  ...;
}

// good
function urlUpdater() {
  let innerUrl = 'https://example.com';
  const DEFAULT = Symbol('DEFAULT');
  return function getOrUpdate(newUrl = DEFAULT) {
    if (newUrl === DEFAULT) {
        return innerUrl;
    }
    // defined in function foo and mutated in getOrUpdate
    innerUrl = newUrl;
  }
}

// 'good'
function isPromiseZalgo() {
  let canary = true;
  // canary accessed in arrow. Changes behavior if using https://github.com/krakenjs/zalgo-promise
  const isZalgoProm = Promise.resolve().then(v => canary);
  canary = false;
  return isZalgoProm;
}

// 'good'
function countIf(vals, fn) {
    let count = 0;
    for (const val of vals) {
      if fn(vals) {
        // defined in countIf, mutated in nested for loop.
        count = count + 1;
      }
    }
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**
Would be a sensible enough to add an option to prefer const.

**Comment by graingert** [comment]
An example from the eslint codebase:

https://github.com/eslint/eslint/blob/0e09973ba6680bff24ebaf94e83ed1db31d1c1b8/lib/util/fix-tracker.js#L88-L106

could be:
```js
function replaceTextRange(range, text) {
        const actualRange = this.retainedRange ?
          [
            Math.min(this.retainedRange[0], range[0]),
            Math.max(this.retainedRange[1], range[1])
          ] :
          range;

        return this.fixer.replaceTextRange(
            actualRange,
            this.sourceCode.text.slice(actualRange[0], range[0]) +
                text +
                this.sourceCode.text.slice(range[1], actualRange[1])
        );
}
```

**Comment by not-an-aardvark** [comment]
Thanks for the proposal. It looks like there are several different cases you're describing here:

1. Replacing simple conditional assignments with ternary expressions

    ```js
    let foo;
    if (condition()) {
      foo = firstValue;
    } else {
      foo = secondValue;
    }
    
    // replaced with:
    
    const foo = condition ? firstValue : secondValue;
    ```

1. Replacing complex one-time assignments with IIFEs

    ```js
    let itWorks;
    try {
      checkIfItWorks();
      itWorks = true;
    } catch (e) {
      itWorks = false;
    }

    // replaced with:

    const itWorks = (() => {
      try {
        checkIfItWorks();
        return true;
      } catch (e) {
        return false;
      }
    })();
    ```

1. Replacing self-references with the `||` operator

    ```js
    let result = foo();
    if (!result) {
      result = bar();
    }
    
    // replaced with:

    const result = foo() || bar();
    ```

I have a few concerns about updating `prefer-const` for these cases:

* This seem out-of-scope for the `prefer-const` rule. I think the `prefer-const` rule is only supposed to flag uses of `let` that could be transformed to `const` without any control flow modification.

    In general, it would probably be possible to remove all usage of `let` with a sufficiently advanced code transform, so in some sense *every* `let` is "unnecessary". However, I don't think it would make sense for `prefer-const` to report every usage of `let`. If you want to disallow all usage of `let`, you could configure the `no-restricted-syntax` rule to disallow it.
* The particular patterns you've described here might not work well when assigning multiple variables for the same condition. In the example below, using ternary expressions would require evaluating `condition()` twice.

    ```js
    let foo;
    let bar;

    if (condition()) {
      foo = 1;
      bar = 2;
    } else {
      foo = 3;
      bar = 4;
    }
    ```

**Comment by graingert** [comment]
Maybe a new rule, rather than a flag to prefer const? I don't think it would be an auto fix rule

**Comment by not-an-aardvark** [comment]
Could you clarify what the rule would do in cases like this?

```js
let foo;
let bar;

if (condition()) {
  foo = 1;
  bar = 2;
} else {
  foo = 3;
  bar = 4;
}
```

At the moment I don't think this preference is common enough to be worth adding to ESLint core, but I'm open to being convinced otherwise. (Of course, you could always enforce this preference yourself by implementing it as a rule in a plugin.)

**Comment by graingert** [comment]
I'm not suggesting an autofixer, I think that would be out of scope. In terms of what I'd prefer for your example, I'd go for:

```js
const mod = condition() ? 0 : 3;
const foo = 1 + mod;
const bar = 2 + mod;
```

**Comment by platinumazure** [comment]
Good grief, how is that *remotely* more readable/better than the example?

**Comment by graingert** [comment]
@platinumazure it expresses the meaning behind the code, condition is presumably some case where we want foo and bar to be 2 bigger or not.

**Comment by not-an-aardvark** [comment]
To clarify, the intent of the example was that `foo` and `bar` would be unrelated. Here is a better example:

```js
let foo;
let bar;

if (condition()) {
  foo = 'the';
  bar = [53, 2, 19];
else {
  foo =  'and';
  bar = [1, 2, 3];
}
```

**Comment by graingert** [comment]
```js
const useWord = condition();
const foo = useWord ? 'the' : [53, 2, 19];
const bar = useWord ? 'and' : [1, 2, 3];
```

**Comment by graingert** [comment]
```js
const isConnective = !condition();
const foo = isConnective ? 'and' : 'the';
const bar = isConnective ? [1, 2, 3] : [53, 2, 19];
```

> the intent of the example was that foo and bar would be unrelated

how often do you set two unrelated values based on the same condition?

**Comment by graingert** [comment]
I've got stuff like:

```js
if (process.env.NODE_ENV === 'production') {
    return { host: 'example.com', port: 443 };
}

return { host: 'localhost', port: 8080 };
```

**Comment by not-an-aardvark** [comment]
> how often do you set two unrelated values based on the same condition?

I've found that it's pretty common ([example](https://github.com/eslint/eslint/blob/d7610f57abd304b318d1e7d2ec20e9d58fb70cbc/lib/rules/indent.js#L643-L649))

**Comment by graingert** [comment]
```js
const nSpaces = context.options[0];
const isTab = nSpaces === "tab";
const indentSize === isTab ? 1 : nSpaces;
const indentType === isTab ? "tab" : "space";

**Comment by not-an-aardvark** [comment]
To be clear, I definitely agree that what you've described is another way that the code could be written. I just don't think the preference is common enough to be worth adding a core rule for.

**Comment by graingert** [comment]
>  I definitely agree that what you've described is another way that the code could be written.

that's extremely unbiased of you

**Comment by graingert** [comment]
> I just don't think the preference is common enough to be worth adding a core rule for.

Guess I'll try peddling a plugin and coax everyone into using it.

**Comment by not-an-aardvark** [comment]
It seems like the team is generally opposed to adding this to core, so I'm going to close this issue. Feel free to link the plugin if you create it!
